### ISSUE TITLE
Rules for nested CSS following css-nesting-1

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#6421]
### What is the problem you're trying to solve?

`css-nesting-1` is being implemented in Chrome. The specification isn't final yet and it might change in the next few weeks.

There are some issues where linters might help authors write correctly nested CSS.

```css
.foo {
  bar & {} /* ambiguous nesting */
}
```

```css
::before {
  bar & {} /* pseudo elements are not allowed */
}
```

At the moment we still need to wait for the final specification for all issues to surface.

### What solution would you like to see?

Some stylelint rules that warn authors when they write nested selectors that are invalid or ambiguous.

Is there interest to add this to stylelint?

**Comment by jeddy3** [comment]
@romainmenke Thanks for the request and for using the template.

> Is there interest to add this to stylelint?

Yes, absolutely! I'm glad you created an issue. I've been meaning to open one myself since the discussion about the nesting spec in https://github.com/stylelint/stylelint/issues/6294#issuecomment-1236131389.

In that issue, we touched on these two nesting patterns:

```css
/* this is valid, but probably unintentional */
{
  && {}
}
```

```css
/* these are equivalent */
a {
  @media (orientation: landscape) { color: red; }
}

a {
  @media (orientation: landscape) { & { color: red; } }
}
```

> Some stylelint rules that warn authors when they write nested selectors that are invalid or ambiguous.

Rules that warn about invalid or ambiguous patterns are a perfect fit for Stylelint. We should eventually add rules to catch:

```css
.foo {
  bar & {} /* ambiguous nesting */
}
```

And

```css
::before {
  bar & {} /* pseudo elements are not allowed */
}
```

The patterns above could fit into one of our existing or upcoming categories:

 - [`no-invalid`](https://stylelint.io/user-guide/rules#invalid) - disallow invalid syntax
 - `no-unmatchable` - disallow valid but unmatchable things, e.g.:
     - https://github.com/stylelint/stylelint/issues/5907
     - https://github.com/stylelint/stylelint/issues/5575
- [`notation`](https://stylelint.io/user-guide/rules#notation) - enforce one representation of things that have multiple 

Or we may need a new category or two, e.g. `no-ambiguous`.

>  The specification isn't final yet and it might change in the next few weeks.

I suggest any rules are first [created as community plugins](https://stylelint.io/developer-guide/plugins) (perhaps bundled into a plugin pack) while the spec stabilises and is adopted by browsers. It's easier to experiment and make changes in plugins than it is to built-in rules.

The closer the plugins follow [our conventions for rules](https://stylelint.io/developer-guide/rules) and [criteria for inclusion](https://stylelint.io/user-guide/rules/) the easier it'll be to port them to built-in rules when the time is right.

I'll label this issue as a discussion so that it can be a place to chat about:
- the names, options and capabilities of the plugins
- what other patterns, in addition to the 4 above, that we'd like to lint
- progress of any plugins
- when the time is right to add built-in rules

Everyone's welcome to chime in. There are opportunities for us to help people write good and consistent nesting CSS.

**Comment by romainmenke** [comment]
Another case :

```css
div {
  .foo, textarea { … }
}
```

This is fully valid but mutations can easily make it invalid.

- splitting `.foo` and `textarea` into two rules -> `textarea {}` is invalid
- removing `.foo` -> `textarea {}` is invalid
- re-order `.foo, textarea` -> `textarea, .foo {}` is invalid

Some authors might prefer to enforce this style:

```css
div {
  .foo, :is(textarea) { … }
}
```

**Comment by sesse** [comment]
Rather than enforcing `:is(textarea)`, why not `& textarea`?

**Comment by romainmenke** [comment]
> Rather than enforcing :is(textarea), why not & textarea?

Updated, I typed this out too quickly.
There are too separate cases, one requires `&<space>`, the other `:is(<type>)`

**Comment by jeddy3** [comment]
[Article from Webkit on nesting](https://webkit.org/blog/13813/try-css-nesting-today-in-safari-technology-preview/).

It have an example of https://github.com/stylelint/stylelint/issues/6421#issuecomment-1304443419 in it:

```css
ul {
  padding-left: 1em;
  :is(article) & {
    padding-left: 0;
  }
}
```

**Comment by osmestad** [comment]
Not a solution for everything, but the ambiguous nesting case I think it can be blocked with this: 
```
'selector-nested-pattern': '^\\W',
```
(In other words, not allowing nested selectors to start with a letter)

**Comment by jeddy3** [comment]
In line with https://github.com/stylelint/stylelint/issues/7396#issue-2035832595 and https://github.com/stylelint/stylelint/issues/6593#issuecomment-2470778938, I'm going to create some issue for us to discuss some of the individual rules as nesting seems to have stabilised and is widely supported.

**Comment by Mouvedia** [comment]
A rule for [this case](https://github.com/stylelint/stylelint/issues/6421#issuecomment-1304286037) would be very useful.

I am not sure about the name though.
`"selector-list-safe-nesting": true`?

**Comment by romainmenke** [comment]
> A rule for https://github.com/stylelint/stylelint/issues/6421#issuecomment-1304286037 would be very useful.

> I am not sure about the name though.
selector-list-safe-nesting?

That is no longer needed.
The nesting syntax was relaxed so that issue no longer exists.

**Comment by romainmenke** [comment]
A rule that might handy is for those cases where there is a `&` in the token stream of the selector while the actual selector is invalid:
- `[&]`: `&` can't be used as the attribute key
- `&__foo`, `&a`: `&` can't be used before tag selectors, it [must appear after](https://codepen.io/romainmenke/pen/MYgwbWY)
